src/main/java/frc/robot/systems/DriveFSMSystem.java

```diff
@@ -0,0 +1,337 @@
+	private double prevEncoderPosArc = 0;
+	private double robotXPosArc = 0;
+	private double robotYPosArc = 0;
+	private double prevGyroAngle = 0;
+
+	// Hardware devices should be owned by one and only one system. They must
+	// be private to their owner system and may not be used elsewhere.
+
+	private CANSparkMax frontRightMotor;
+	private CANSparkMax backRightMotor;
+	private CANSparkMax frontLeftMotor;
+	private CANSparkMax backLeftMotor;
+
+	private AHRS gyro;
+
+	/* ======================== Constructor ======================== */
+	/**
+	 * Create FSMSystem and initialize to starting state. Also perform any
+	 * one-time initialization or configuration of hardware required. Note
+	 * the constructor is called only once when the robot boots.
+	 */
+	public DriveFSMSystem() {
+		// Perform hardware init
+
+		frontRightMotor = new CANSparkMax(HardwareMap.CAN_ID_SPARK_DRIVE_FRONT_RIGHT,
+											CANSparkMax.MotorType.kBrushless);
+		frontLeftMotor = new CANSparkMax(HardwareMap.CAN_ID_SPARK_DRIVE_FRONT_LEFT,
+											CANSparkMax.MotorType.kBrushless);
+		backRightMotor = new CANSparkMax(HardwareMap.CAN_ID_SPARK_DRIVE_BACK_RIGHT,
+											CANSparkMax.MotorType.kBrushless);
+		backLeftMotor = new CANSparkMax(HardwareMap.CAN_ID_SPARK_DRIVE_BACK_LEFT,
+											CANSparkMax.MotorType.kBrushless);
+
+		frontRightMotor.getEncoder().setPosition(0);
+		frontLeftMotor.getEncoder().setPosition(0);
+		backRightMotor.getEncoder().setPosition(0);
+		backLeftMotor.getEncoder().setPosition(0);
+
+		finishedMovingStraight = false;
+		finishedTurning = false;
+
+		// Reset state machine
+		reset();
+	}
+
+	/* ======================== Public methods ======================== */
+	/**
+	 * Return current FSM state.
+	 * @return Current FSM state
+	 */
+	public FSMState getCurrentState() {
+		return currentState;
+	}
+	/**
+	 * Reset this system to its start state. This may be called from mode init
+	 * when the robot is enabled.
+	 *
+	 * Note this is distinct from the one-time initialization in the constructor
+	 * as it may be called multiple times in a boot cycle,
+	 * Ex. if the robot is enabled, disabled, then reenabled.
+	 */
+	public void reset() {
+		currentState = FSMState.TELEOP_STATE;
+
+		// Call one tick of update to ensure outputs reflect start state
+		update(null);
+	}
+	/**
+	 * Update FSM based on new inputs. This function only calls the FSM state
+	 * specific handlers.
+	 * @param input Global TeleopInput if robot in teleop mode or null if
+	 *        the robot is in autonomous mode.
+	 */
+	public void update(TeleopInput input) {
+		rawGyroAngle = gyro.getAngle();
+		updateLineOdometry();
+		updateArcOdometry();
+		switch (currentState) {
+			case START_STATE:
+				handleStartState(input);
+				break;
+
+			case TELEOP_STATE:
+				handleTeleOpState(input);
+				break;
+
+			case FORWARD_STATE_10_IN:
+				handleForwardOrBackwardState(input, 10,
+					forwardStateInitialEncoderPos != -1 ? forwardStateInitialEncoderPos
```

Comment:
What if we just set forwardStateInitialEncoderPos to the current encoder position in handleTurnState and handleStartState? Or even better, what if we just call `frontLeftMotor.getEncoder().setPosition(0);`

Reply:
The encoder position cannot be reset, because that would mess with the odometry calculations. Instead, I simply passed forwardStateInitialEncoderPos as the parameter and within the handleForwardState method I set it to the current encoder position if forwardStateInitialEncoderPos = -1.